src/systems/implicit_system.C

```diff
@@ -225,6 +226,21 @@ SparseMatrix<Number> & ImplicitSystem::add_matrix (const std::string & mat_name)
 }
 
 
+void ImplicitSystem::remove_matrix (const std::string & mat_name)
+{
+  //Return if the matrix does not exist
+  if (!(this->have_matrix(mat_name)))
+    return;
+
+  matrices_iterator pos = _matrices.find (mat_name);
+
+  pos->second->clear();
+  delete pos->second;
+
+  _matrices.erase(mat_name);
+}
```

Comment:
But this still does 3 lookups: 1 for `have_matrix()`, one for `find()` and one for `erase(key)`. Also I don't think it's necessary to call `clear()` followed by delete because all the `SparseMatrix` destructors call `clear()`. So, the following should work:
```
matrices_iterator pos = _matrices.find (mat_name);
if (pos==end) return;
delete pos->second;
_matrices.erase(pos);
```